slasher/rpc/detect_update_min_max_span.go

```diff
@@ -0,0 +1,115 @@
+	targetEpoch, span, spanMap, err := ss.detectSlashingByEpochSpan(source, target, validatorIdx, detectEpochMaxSpan)
+	if err != nil {
+		return 0, err
+	}
+	if targetEpoch > 0 {
+		return targetEpoch, nil
+	}
+	for i := uint64(1); i < target-source; i++ {
+		val := uint32(span - i - 1)
+		if _, ok := spanMap.EpochSpanMap[source+i]; !ok {
+			spanMap.EpochSpanMap[source+i] = &ethpb.MinMaxEpochSpan{}
+		}
+		if spanMap.EpochSpanMap[source+i].MaxEpochSpan < val {
+			spanMap.EpochSpanMap[source+i].MaxEpochSpan = val
+		} else {
+			break
+		}
+	}
+	if err := ss.SlasherDB.SaveValidatorSpansMap(validatorIdx, spanMap); err != nil {
+		return 0, err
+	}
+	return 0, nil
+}
+
+// DetectAndUpdateMinEpochSpan is used to detect surround and update the min epoch span
+// of an incoming attestation.
+// min span is the span between the current attestation and the closest attestation target
+// distance.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMinEpochSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (uint64, error) {
+	targetEpoch, _, spanMap, err := ss.detectSlashingByEpochSpan(source, target, validatorIdx, detectEpochMinSpan)
+	if err != nil {
+		return 0, err
+	}
+	if targetEpoch > 0 {
+		return targetEpoch, nil
+	}
+	if source == 0 {
+		return 0, nil
+	}
+	for i := source - 1; i > 0; i-- {
+		val := uint32(target - (i))
+		if _, ok := spanMap.EpochSpanMap[i]; !ok {
+			spanMap.EpochSpanMap[i] = &ethpb.MinMaxEpochSpan{}
+		}
+		if spanMap.EpochSpanMap[i].MinEpochSpan == 0 || spanMap.EpochSpanMap[i].MinEpochSpan > val {
+			spanMap.EpochSpanMap[i].MinEpochSpan = val
+		} else {
+			break
+		}
+	}
+	if err := ss.SlasherDB.SaveValidatorSpansMap(validatorIdx, spanMap); err != nil {
+		return 0, errors.Wrap(err, "could not save validator spans")
+	}
+	return 0, nil
+}
+
+func detectEpochMaxSpan(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
+	maxSpan := uint64(recorderEpochSpans.MaxEpochSpan)
+	if maxSpan > attestationEpochSpan {
+		return maxSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+func detectEpochMinSpan(attestationEpochSpan uint64, recorderEpochSpans *ethpb.MinMaxEpochSpan, attestationSourceEpoch uint64) uint64 {
+	minSpan := uint64(recorderEpochSpans.MinEpochSpan)
+	if minSpan < attestationEpochSpan {
+		return minSpan + attestationSourceEpoch
+	}
+	return 0
+}
+
+// detectSlashingByEpochSpan is used to detect if a slashable event is present
+// in the db by checking either the closest attestation target or the furthest
+// attestation target. this method receives the detection func in order to be
+// compatible for both cases.
+func (ss *Server) detectSlashingByEpochSpan(source, target, validatorIdx uint64, detectionFunc detect) (uint64, uint64, *ethpb.EpochSpanMap, error) {
```

Comment:
See comment from earlier, I'd recommend passing in a detector interface into this func and having structs that implement such interface